src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    internal class IpcTransport
+    {
+        private string _ipcTransportPath = null;
+        private int? _pid = null;
+
+        private static double ConnectTimeoutMilliseconds { get; } = TimeSpan.FromSeconds(3).TotalMilliseconds;
+        public static string IpcRootPath { get; } = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"\\.\pipe\" : Path.GetTempPath();
+        public static string DiagnosticsPortPattern { get; } = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? @"^dotnet-diagnostic-(\d+)$" : @"^dotnet-diagnostic-(\d+)-(\d+)-socket$";
+
+        /// <summary>
+        /// Creates a reference to a .NET process's IPC Transport
+        /// by getting a reference to the Named Pipe or Socket
+        /// specified in ipcTransportPath
+        /// </summary>
+        /// <param name="ipcTransportPath">The fully qualified path, including filename, of the IPC Transport</param>
+        /// <returns>A reference to the IPC Transport</returns>
+        public IpcTransport(string ipcTransportPath)
+        {
+            _ipcTransportPath = ipcTransportPath;
+        }
+
+        /// <summary>
+        /// Creates a reference to a .NET process's IPC Transport
+        /// using the default rules for a given pid
+        /// </summary>
+        /// <param name="pid">The pid of the target process</param>
+        /// <returns>A reference to the IPC Transport</returns>
+        public IpcTransport(int pid)
+        {
+            _pid = pid;
+        }
+
+        /// <summary>
+        /// Connects to the underlying IPC Transport and opens a read/write-able Stream
+        /// </summary>
+        /// <returns>A Stream for writing and reading data to and from the target .NET process</returns>
+        public Stream Connect()
+        {
+            if (!string.IsNullOrEmpty(_ipcTransportPath))
+            {
+                return ConnectViaPath();
+            }
+            else if (_pid.HasValue)
+            {
+                return ConnectViaPid();
+            }
+            else
+            {
+                throw new ApplicationException("Cannot connect to Diagnostics IPC Transport without a PID or specific path");
```

Comment:
Instead of throwing an `ApplicationException` can we throw some subclass of DiagnosticsClientException?